base/cvd/cuttlefish/host/libs/zip/lazily_loaded_file.cc

```diff
@@ -0,0 +1,159 @@
+
+  impl->contents_file_ = SharedFD::Open(filename, O_CREAT | O_RDWR, 0644);
+  impl->filename_ = std::move(filename);
+  impl->callback_ = std::move(callback);
+  impl->seek_pos_ = 0;
+
+  CF_EXPECT(impl->ReadMetadata());
+
+  return LazilyLoadedFile(std::move(impl));
+}
+
+LazilyLoadedFile::LazilyLoadedFile(std::unique_ptr<Impl> impl)
+    : impl_(std::move(impl)) {}
+
+LazilyLoadedFile::LazilyLoadedFile(LazilyLoadedFile&& other) {
+  std::swap(impl_, other.impl_);
+}
+
+LazilyLoadedFile::~LazilyLoadedFile() {
+  if (!impl_) {
+    return;
+  }
+  Result<void> res = impl_->WriteMetadata();
+  if (!res.ok()) {
+    LOG(WARNING) << "fragment update failure: " << res.error().FormatForEnv();
+  }
+}
+
+LazilyLoadedFile& LazilyLoadedFile::operator=(LazilyLoadedFile&& other) {
+  impl_.reset();
+  std::swap(impl_, other.impl_);
+  return *this;
+}
+
+Result<size_t> LazilyLoadedFile::Read(char* data, size_t size) {
+  CF_EXPECT(impl_.get());
+  return CF_EXPECT(impl_->Read(data, size));
+}
+
+Result<void> LazilyLoadedFile::Seek(size_t location) {
+  CF_EXPECT(impl_.get());
+  impl_->seek_pos_ = location;
+  return {};
+}
+
+std::string LazilyLoadedFile::Impl::MetadataFile() const {
+  return filename_ + ".frag_data";
+}
+
+Result<void> LazilyLoadedFile::Impl::ReadMetadata() {
+  SharedFD metadata_fd = SharedFD::Open(MetadataFile(), O_CREAT | O_RDWR, 0644);
+  CF_EXPECTF(metadata_fd->IsOpen(), "Failed to open {}: {}", MetadataFile(),
+             metadata_fd->StrError());
+
+  std::string data;
+  CF_EXPECT_GE(ReadAll(metadata_fd, &data), 0, metadata_fd->StrError());
+
+  Result<DisjointRangeSet> parsed_res = DeserializeDisjointRangeSet(data);
+  if (parsed_res.ok()) {
+    already_downloaded_ = std::move(*parsed_res);
+  } else {
+    LOG(WARNING) << "Invalid fragments: " << parsed_res.error().FormatForEnv();
+  }
+
+  return {};
+}
+
+Result<void> LazilyLoadedFile::Impl::WriteMetadata() {
+  std::string new_metadata_name = MetadataFile() + ".XXXXXX";
+  SharedFD new_metadata = SharedFD::Mkstemp(&new_metadata_name);
+  CF_EXPECT(new_metadata->IsOpen(), new_metadata->StrError());
+  CF_EXPECT(new_metadata->Chmod(0644));
+
+  std::string data = Serialize(already_downloaded_);
+  CF_EXPECT_EQ(WriteAll(new_metadata, data), data.size(),
+               new_metadata->StrError());
+
+  CF_EXPECT(RenameFile(new_metadata_name, MetadataFile()));
+
+  return {};
+}
+
+Result<size_t> LazilyLoadedFile::Impl::Read(char* data, size_t size) {
+  // NOLINTNEXTLINE(misc-include-cleaner): SEEK_SET
+  CF_EXPECT_EQ(contents_file_->LSeek(seek_pos_, SEEK_SET), seek_pos_,
+               contents_file_->StrError());
+  if (already_downloaded_.ContainsRange(seek_pos_, seek_pos_ + size)) {
+    ssize_t data_read = contents_file_->Read(data, size);
+    CF_EXPECT_GE(data_read, 0, contents_file_->StrError());
+    seek_pos_ += data_read;
+    return data_read;
+  }
+  CF_EXPECT(callback_->Seek(seek_pos_));
+  size_t data_read = CF_EXPECT(callback_->Read(data, size));
```

Comment:
This can potentially read again a portion that has already been read if we already have part of the requested range.

To avoid multiple reads of the same range we could add functionality to `DisjointRangeSet` to return the first start and end of range it has at or after a given position. With that in hand we can request more data if we don't have the beginning of the range to read and then return as much data as we have, but possibly less than the caller asked for.

For example, let's say we have [3, 5) and [7, 10) and we're asked to read 10 bytes at position 0. Then we read with the callback [0, 3) and return to the user [0,5). They want 10 bytes, so they read again 5 bytes at position 5, we download [5, 7) and finally return the bytes from [5, 10).

Reply:
This still may download the same data more than once. In the example I originally provided it will download the entire [0, 10) range even though it already has parts of it.

This minimizes the number of calls to the read callback as opposed to how much data is gotten from it. If that's the objective, please add a comment that mentions it.

Reply:
The objective is to minimize round trips against the IO source rather than minimizing bandwidth.

Reply:
Added a comment.